src/Host/Client/Impl/Host/RHost.cs

```diff
@@ -275,7 +288,15 @@ public sealed partial class RHost : IDisposable, IRExpressionEvaluator {
             } else {
                 result = new REvaluationResult(response[2], error, parseStatus);
             }
-            request.CompletionSource.SetResult(result);
+            request.CompletionSource.TrySetResult(result);
+        }
+
+        private void CancelRemainingEvaluationRequests() {
+            var requests = _evalRequests.Values;
+            _evalRequests.Clear();
```

Comment:
There is no harm if Clean will be called concurrently with request processing, but there can be a problem if new request is added after `Clear` is called. I'll look into it.

---
In reply to: [71067479](https://github.com/Microsoft/RTVS/pull/2080#discussion_r71067479) [](ancestors = 71067479)